### PR TITLE
feat(opentelemetry-node): add cloud/container resource detectors

### LINKED ISSUE #128: add cloud/container resource detectors by default
Add the available cloud/container resource detectors by default in ElasticNodeSDK:

```js
        const defaultConfig = {
            resourceDetectors: [
...
```

- One consideration might be to see about making them silent if they fail to detect resources. We don't want the noise if we have all these on by default.
- Do we want an `ELASTIC_OTEL_` config option to turn those off a la elastic-apm-node's `ELASTIC_APM_CLOUD_PROVIDER`? Something that would be easier than having to come up with a working subset for `OTEL_NODE_RESOURCE_DETECTORS`?
- If a user attempts to use `OTEL_NODE_RESOURCE_DETECTORS`, is there a value that will still get our Elastic detector that adds the distro resource keys?

### PR DESCRIPTION
**Pull request by david-luna**
Closes #128 

This PR adds the container and the cloud resource detectors (alibaba, aws, azure & gcp). By default all are included unless the env var `OTEL_NODE_RESOURCE_DETECTORS` is set. I that case the distro includes according the value set in the environment. Some highlights:

- resolution of which detectors is moved to a separate module instead like it was done for instrumentations
- env var affects to **all detectors**. So if uses sets a list of detectors our `distro` detector may be excluded.
- looking at `@opentelemetry/auto-instrumentations-node` there are a couple of detectors not related to cloud/containers left to add

**Question**
This detectors resolution gives more priority to setup via code rather than the environment setup via environment vars. I wonder if that's the behaviour we want 🤔 

Vanilla SDK seems to have this behaviour too ([link](https://github.com/open-telemetry/opentelemetry-js/blob/6481396a6b476310356ed276e17113658e050470/experimental/packages/opentelemetry-sdk-node/src/sdk.ts#L131)) and its `@opentelemetry/auto-instrumentations-node` the one looking up to env var `OTEL_NODE_RESOURCE_DETECTORS`

### CODE REVIEW

**Comment by david-luna** [comment]
Tests are failing since some detectors generate spans. An example is HTTP spans being generated by GCP detector

```
node -r @elastic/opentelemetry-node ./test/fixtures/use-http-server.js

------ trace 350a53 (2 spans) ------
       span 1edd33 "GET" (10.9ms, STATUS_CODE_ERROR, SPAN_KIND_CLIENT, GET http://127.0.0.1:56695/ -> 404)
  +6ms `- span 059043 "GET" (1.9ms, SPAN_KIND_SERVER, GET http://127.0.0.1:56695/ -> 404)
------ trace 68c04e (1 span) ------
       span 190981 "GET" (6.3ms, STATUS_CODE_ERROR, SPAN_KIND_CLIENT, GET http://metadata.google.internal./computeMetadata/v1/instance)
------ trace f32a41 (2 spans) ------
       span 8af26e "POST" (3.3ms, SPAN_KIND_CLIENT, POST http://127.0.0.1:56695/echo -> 200)
  +2ms `- span 2df652 "POST" (0.8ms, SPAN_KIND_SERVER, POST http://127.0.0.1:56695/echo -> 200)
------ trace 3230c3 (1 span) ------
       span b94337 "GET" (3004.0ms, STATUS_CODE_ERROR, SPAN_KIND_CLIENT, GET http://169.254.169.254/computeMetadata/v1/instance)
```

**Comment by david-luna** [comment]
As discussed in today's meeting with @trentm 
- distro resource detector is added always
- check auto-instrumentations if GCP spans are also created and create an issue for that
- filter out the GCP spans from the `sortedSpans` getter in tests

**Comment by trentm** [comment]
@david-luna This is still draft. I wasn't sure if this was ready for review yet.

**Comment by david-luna** [comment]
> @david-luna This is still draft. I wasn't sure if this was ready for review yet.

There was a linting issue happening in CI but not in local so it wasn't ready yet. Last commit contains a fix for `gen:types` npm script.

**Comment by david-luna** [comment]
Note on the lint failure.

```
npm ls --all

...
 ├─┬ @opentelemetry/resource-detector-gcp@0.29.9
  │ ├── @opentelemetry/api@1.8.0 deduped
  │ ├── @opentelemetry/core@1.24.1 deduped
  │ ├── @opentelemetry/resources@1.24.1 deduped
  │ ├── @opentelemetry/semantic-conventions@1.24.1 deduped
  │ └─┬ gcp-metadata@6.1.0
  │   ├─┬ gaxios@6.6.0
  │   │ ├── extend@3.0.2
  │   │ ├─┬ https-proxy-agent@7.0.4
  │   │ │ ├─┬ agent-base@7.1.1
  │   │ │ │ └── debug@4.3.4 deduped
  │   │ │ └── debug@4.3.4 deduped
  │   │ ├── is-stream@2.0.1 deduped
  │   │ ├─┬ node-fetch@2.7.0
  │   │ │ ├── UNMET OPTIONAL DEPENDENCY encoding@^0.1.0
  │   │ │ └─┬ whatwg-url@5.0.0
  │   │ │   ├── tr46@0.0.3.       <----- THIS VERSION OF `tr46` HAS NOT LICENSE FILE
  │   │ │   └── webidl-conversions@3.0.1
  │   │ └── uuid@9.0.1
  │   └─┬ json-bigint@1.0.0
  │     └── bignumber.js@9.1.2
 ...
```

Linting if failing because of a missing license file, was added in https://github.com/jsdom/tr46/tree/v1.0.0. The `package.json` file has the field "license" set correctly but I'm not sure if we should add a fallback to that field if the file is missing.

Any thoughts @trentm ?

**Comment by trentm** [comment]
> The `package.json` file has the field "license" set correctly but I'm not sure if we should add a fallback to that field if the file is missing.

I think it is fine to add a fallback in the lint script for this.  The package.json "license" field says MIT. The very next commit on the tr46 repo adds the license file that is then included in subsequent published versions.

I think it would be fine to try a PR on node-fetch to update their whatwg-url dep, but I don't think that needs to block us here.

**Comment by david-luna** [comment]
@trentm I just realized `./script/gen-notice.sh` already has a fallback mechanism for packages that do not have a license file. I've added a new entry for this fallback in https://github.com/elastic/elastic-otel-node/pull/214/commits/6888419d1dba1f7e36c8560fbe3806529f34d864

File: packages/opentelemetry-node/lib/detectors.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        process.env['OTEL_NODE_RESOURCE_DETECTORS'] || 'all';
+
+    if (detectorsFromEnv === 'none') {
+        return [];
+    }
+
+    const detectorKeys =
+        detectorsFromEnv === 'all'
+            ? Object.keys(defaultDetectors)
+            : detectorsFromEnv.split(',');
```
Comment (trentm) [inline packages/opentelemetry-node/lib/detectors.js:114]: The logic is *slightly* difference in auto-instr-node: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/dadf30828e7008df7b85c8bd7eaacb90770aac5c/metapackages/auto-instrumentations-node/src/utils.ts#L235-L244

E.g. `OTEL_NODE_RESOURCE_DETECTORS=host,os,none`  with a "none" in there the behaviour differs. Also if 'all' is *one* of the values in there.

Personally, I think it would be nice to `.trim()` each value so `OTEL_NODE_RESOURCE_DETECTORS=host, os` works. `auto-instr-node` does trim in its handling of `process.env.OTEL_NODE_ENABLED_INSTRUMENTATIONS`, FWIW.

File: packages/opentelemetry-node/lib/detectors.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    /** @type {Array<DetectorSync | DetectorSync[]>} */
+    const resolvedDetectors = [distroDetectorSync];
+
+    for (const key of detectorKeys) {
+        if (defaultDetectors[key]) {
+            resolvedDetectors.push(defaultDetectors[key]);
+        } else {
+            log.debug(
```
Comment (trentm) [inline packages/opentelemetry-node/lib/detectors.js:122]:
```suggestion
            log.warn(
```

File: packages/opentelemetry-node/test/instr-http.test.js [hunk 3]
```diff
@@ -76,6 +76,7 @@ const testFixtures = [
             //     ------ trace b8467d (2 spans) ------
             //            span bc8a2c "POST" (3.8ms, SPAN_KIND_CLIENT, POST http://127.0.0.1:64972/echo -> 200)
             //      +2ms `- span 4e7adf "POST" (1.1ms, SPAN_KIND_SERVER, POST http://127.0.0.1:64972/echo -> 200)
+            // const spans = col.sortedSpans.slice(RESOURCE_DETECTOR_SPAN_COUNT);
```
Comment (trentm) [inline packages/opentelemetry-node/test/instr-http.test.js:79]: debugging code?

File: packages/opentelemetry-node/test/testutils.js [hunk 4]
```diff
@@ -332,6 +336,16 @@ class TestCollector {
 
             return aStartInt < bStartInt ? -1 : aStartInt > bStartInt ? 1 : 0;
         });
+
+        // NOTE: some resource detectors are starting spans and this affects
+        // the tests we have implmented until now (2024-06)
+        // - internal components like detector shouldn't be crating traces
```
Comment (trentm) [inline packages/opentelemetry-node/test/testutils.js:342]:
```suggestion
        // NOTE: Ignore spans from the GCP detector for testing. It shouldn't be
        // creating spans, but that should be fixed upstream.
```

Perhaps an internal issue we can track to know when this hack can be removed?

File: packages/opentelemetry-node/test/instr-http.test.js (hunk 3 above)
Comment (david-luna) [reply packages/opentelemetry-node/test/instr-http.test.js:79]: yup, I've missed that

File: packages/opentelemetry-node/lib/detectors.js (hunk 1 above)
Comment (david-luna) [reply packages/opentelemetry-node/lib/detectors.js]: I've applied your suggestions. Thanks :)

File: packages/opentelemetry-node/test/testutils.js (hunk 4 above)
Comment (david-luna) [reply packages/opentelemetry-node/test/testutils.js:342]: https://github.com/elastic/elastic-otel-node/issues/222

**Comment by david-luna** [comment]
@trentm I think we can add some tests in a similar way than https://github.com/elastic/elastic-otel-node/pull/220

WDYT?

File: packages/opentelemetry-node/lib/elastic-node-sdk.js [hunk 5]
```diff
@@ -58,20 +53,11 @@ class ElasticNodeSDK extends NodeSDK {
...
         const defaultConfig = {
-            resourceDetectors: [
-                // Elastic's own detector to add some metadata
-                distroDetectorSync,
-                envDetectorSync,
-                processDetectorSync,
-                // hostDetectorSync is not currently in the OTel default, but may be added
-                hostDetectorSync,
-            ],
+            resourceDetectors: resolveDetectors(opts.resourceDetectors),
```
Comment (trentm) [inline packages/opentelemetry-node/lib/elastic-node-sdk.js:56]: Not for this PR. I wonder if eventually we may want to export something similar to `getInstrumentations()` that enables a user to user our SDK and get the default set of detectors **plus one or more custom detectors**.  Or perhaps this is something that is more generally solved with some concept of SDK plugins or extensions as briefly discussed on the OTel JS SIG.

**Review by trentm — APPROVED** [review]

**Comment by trentm** [comment]
> I'm hesitant if wee need to add more tests for other scenarios like the behaviour when the value `all` or `none` is present in the list. WDYT?

I think we are fine without the tests. Your call if you'd like to add more.

File: packages/opentelemetry-node/lib/elastic-node-sdk.js (hunk 5 above)
Comment (david-luna) [reply packages/opentelemetry-node/lib/elastic-node-sdk.js:56]: I'd say if the number of detectors grow enough we will have a case for that API to be exported. We can discuss it and crate an issue if we agree it's necessary
